### ISSUE TITLE
📎 Improve error reporting

### ISSUE THREAD

**Issue by Conaclos** [biomejs/biome#2730]
### Description

#2659 brought something to light for me: error reporting is uninformative for users and doesn't help them to understand the source of the issue or how to find a workaround.

Biome is a toolchain, a bug can come from any tool, and it is hard to tell without investment of the user.

It could be nice if we could indicate by which tool the error has raised, and even in which unit of work (for example a rule for the linter) and ideally the linted/formatted node (the queried node in the case of the linter).

Taking #2659 as an example.
Here is the current error:

```
Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would
appreciate it if you could report it to
https://github.com/biomejs/biome/issues/ along with the
following information to help us fixing the issue:

Source Location: crates/biome_js_semantic/src/semantic_model/scope.rs:115:33
Thread Name: biome::worker_0
Message: no entry found for key

types.ts internalError/panic  INTERNAL  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

✖ processing panicked: no entry found for key
  
⚠ This diagnostic was derived from an internal Biome error.
  Potential bug, please report it if necessary.
```

And here is an example of a better diagnostic:

```
types.ts internalError/panic  INTERNAL  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

✖ Biome encountered an internal error while linting the file with the
 rule suspicious/noRedeclare on the following code

    1 │
  > 2 │ type Element<T> = T extends Array<infer U> ? U : never
    3 │ 

ℹ This diagnostic was derived from the following internal Biome error:
 Source Location: crates/biome_js_semantic/src/semantic_model/scope.rs:115:33
 Thread Name: biome::worker_0
 Message: no entry found for key

ℹ This is a bug in Biome, not an error in your code, and we would
 appreciate it if you could report it to
 https://github.com/biomejs/biome/issues/

ℹ As a workaround try to disable the lint rule suspicious/noRedeclare
 or ignoring the file `types.ts`.

```

**Comment by Conaclos** [comment]
> Have you checked if it's technically possible to retrieve the information you'd like to see?

No, I haven't checked how to proceed, it's an invitation to investigate. Perhaps I should clarify the issue description.
Because we have a multi-threaded architecture, panics are isolated in threads? I wonder if we could get some data from the main thread.

> The only information we might be able to have, is the stack trace, however if it's not printed here, I doubt we have it.

I was thinking of doing this to get some information such as the currently running linter rule.

**Comment by dyc3** [comment]
We can wrap evaluating rules with [catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html#) to catch panics, and that would allow us to surface the rule that was being evaluated. I'm not so sure about being able to surface the actual user code that resulted in the panic.

**Comment by ematipico** [comment]
We already have that. 

https://github.com/biomejs/biome/blob/main/crates%2Fbiome_cli%2Fsrc%2Fexecute%2Ftraverse.rs#L594

**Comment by dyc3** [comment]
What if we did the `catch_unwind` here, where the rule is being evaluated?

https://github.com/biomejs/biome/blob/5088667ce337d34bc6f7a19887b07b06384d8fdc/crates/biome_analyze/src/registry.rs#L431

This function seems to have all the context we want in the new error messages.
